### PR TITLE
parallel iter element

### PR DESCRIPTION
**Pull request by hunjixin**
parallel iter element

### CODE REVIEW

**Comment by hunjixin** [comment]
@whyrusleeping

File: amt.go [hunk 1]
```diff
@@ -295,6 +306,50 @@ func (n *Node) forEachAt(ctx context.Context, bs cbor.IpldStore, height int, sta
...
+			return errors.New("just for stop")
+		default:
+		}
+		wg.Add(1)
+		select {
+		case <-toProcess:
+			go func() {
+				defer func() {
+					wg.Done()
+				}()
```
Comment (warpfork) [inline amt.go:330]: This can be just `defer wg.Done()` -- there's no need for the anonymous function that's wrapping it.

File: amt.go [hunk 2]
```diff
@@ -295,6 +306,50 @@ func (n *Node) forEachAt(ctx context.Context, bs cbor.IpldStore, height int, sta
...
+	errCh := make(chan struct{})
+	toProcess := make(chan struct{}, parNum)
+	for i := 0; i < parNum; i++ {
+		toProcess <- struct{}{}
+	}
+
+	var wg sync.WaitGroup
+	multiErr := &MultiError{}
+	wrapCh := func(u uint64, deferred *cbg.Deferred) error {
+		select {
```
Comment (warpfork) [inline amt.go:319]: Somewhere in here, I would expect to see a check for context cancellation.

I'm also not sure why there's several `select` blocks.

File: amt.go [hunk 3]
```diff
@@ -295,6 +306,50 @@ func (n *Node) forEachAt(ctx context.Context, bs cbor.IpldStore, height int, sta
 
 }
 
+func (n *Node) parForEachAt(ctx context.Context, bs cbor.IpldStore, height int, start, offset uint64, parNum int, cb func(uint64, *cbg.Deferred) error) error {
+	errCh := make(chan struct{})
+	toProcess := make(chan struct{}, parNum)
+	for i := 0; i < parNum; i++ {
+		toProcess <- struct{}{}
+	}
```
Comment (warpfork) [inline amt.go:314]: I don't understand what this channel is for.  It appears to be a some kind of source token bucket, but... the `wg sync.WaitGroup` further down would appear to also be doing the exact same thing (and is a more appropriate tool for that purpose).

**Comment by warpfork** [comment]
I did some brief code review on this, but am also not the maintainer of this repo, and am not sure what the trajectory for pursuing this work would be.

I'm not sure there's going to be a desire from the maintainers of this repo to merge additional code (which means additional maintenance burden for them) without a very compelling reason.

A PR for parallelization is especially tricky because A) a parallelization strategy really needs careful design; and B) a parallelization system needs careful benchmarking to be sure that it's actually improving things.

**Comment by jennijuju** [comment]
@hunjixin Thank you for contributing, base on @warpfork and @Stebalien 's evaluation, I'm gonna close this pr for now!

**Pull request closed without merging**